app/models/training/module.rb

```diff
@@ -0,0 +1,227 @@
+# USE memcache
+
+require_relative 'content'
+
+module Training
+  class Module < Content
+    # @return [String]
+    def self.content_type_id
+      'trainingModule'
+    end
+
+    # has_many :pages, class_name: 'Training::Page'
+    # has_many :questions, class_name: 'Training::Question'
+    # has_many :videos, class_name: 'Training::Video'
+
+    # @return [Array<Training::Module>]
+    def self.ordered
+      load_children(0).order(:position).load!.to_a
+    end
+
+    # @return [Training::Module]
+    def self.by_id(id)
+      load_children(0).find(id)
+    end
+
+    # @return [Training::Module]
+    def self.by_name(name)
+      load_children(0).find_by(name: name).first
+    end
+
+    # entry references ---------------------------------
+
+    # @example
+    #   mod.thumbnail => "//images.ctfassets.net/dvmeh832nmjc/6ICCjd5b2gVc1jMHwgQHpH/a074dbf76e101efcca35dac2e1de6638/1-1-1-1-869488712.jpg"
+    #
+    # @return [String, nil]
+    def thumbnail_url
+      return '//external-image-resource-placeholder' if fields[:image].blank?
+
+      ContentfulModel::Asset.find(fields[:image].id).url
+    end
+
+    # @return [Training::Module, nil]
+    def depends_on
+      self.class.by_id(fields[:depends_on].id) if fields[:depends_on]
+    end
+
+    # @return [Array<Training::Page, Training::Video, Training::Question>] source of truth for content order
+    def content
+      return [] if draft?
+
+      fields[:pages].map do |child_link|
+        fetch_or_store(child_link.id) { child_by_id(child_link.id) }
+      end
+    end
+
+    # @return [Hash{ Integer=>Array<Module::Content> }]
+    def content_by_submodule
+      content.group_by(&:submodule)
+    end
+
+    def content_by_topic
+      content.group_by { |page|
+        [page.submodule, page.topic] unless page.topic.zero? # or nil? if we remove default to 0
+      }.except(0)
+    end
+
+    # @return [Integer]
+    def topic_count
+      content_by_topic.count
+    end
+
+    # @return [Integer]
+    def submodule_count
+      content_by_submodule.count
+    end
+
+    # Selects from ordered array
+    #
+    # @return [Training::Page, Training::Video, Training::Question]
+    def page_by_id(id)
+      content.find { |page| page.id.eql?(id) }
+    end
+
+    # Selects from ordered array
+    #
+    # @return [Training::Page, Training::Video, Training::Question]
+    def page_by_name(name)
+      content.find { |page| page.name.eql?(name) }
+    end
+
+    # state ---------------------------------
+
+    # @return [Boolean]
+    def draft?
+      fields.fetch(:pages, []).none?
+    end
+
```

Comment:
It is a boolean

Reply:
Sorry, I meant a boolean attribute on contentful.

Reply:
That could well be the better option depending on how the Content team want to work. I just wanted to get the reciprocal links working and avoid adding attributes if I could. For example, they'd either change the boolean attribute on the module, or publish the draft content it contains, both are just as easy in the interface but then it sort of feels like we are competing with the Contentful way of determining state. The later made more sense to me right now.

Reply:
Yes, it would be nice if we had a different word for it -- the state we are considering here is different from what Contentful provides.  You can have a published page and a separate version of that page that is "draft" from a contentful perspective.  From an application perspective it is either a published course or not a published course (unpublished?) and once published it is unlikely or highly suspect if it were to be become unpublished, so a one-way state change.